packages/mdc-list/foundation.js

```diff
@@ -66,16 +67,42 @@ class MDCListFoundation extends MDCFoundation {
    */
   constructor(adapter) {
     super(Object.assign(MDCListFoundation.defaultAdapter, adapter));
-    /** {boolean} */
+    /** @private {boolean} */
     this.wrapFocus_ = false;
-    /** {boolean} */
+
+    /** @private {boolean} */
     this.isVertical_ = true;
-    /** {boolean} */
+
+    /** @private {boolean} */
     this.isSingleSelectionList_ = false;
-    /** {number} */
+
+    /** @private {!Index} */
     this.selectedIndex_ = -1;
-    /** {boolean} */
+
+    /** @private {number} */
+    this.focusedItemIndex_ = -1;
+
+    /** @private {boolean} */
     this.useActivatedClass_ = false;
+
+    /** @private {boolean} */
+    this.toggleCheckbox_ = true;
+
+    /** @private {boolean} */
+    this.isCheckboxList_ = false;
+
+    /** @private {boolean} */
+    this.isRadioList_ = false;
+  }
+
+  init() {
+    if (this.adapter_.getListItemCount() === 0) return;
```

Comment:
For some reason I can't respond to the original comment thread anymore (thanks GitHub?)...

> Regarding lazy list update, I've added `.init()` call to component's `layout()` method. Otherwise, `.init()` is called during initialization anyway?

I know it's called during initialization, my point is that if there are no list items during initialization it's never going to detect the list type. Adding a `foundation.init` call to the component's `layout` method seems like an unorthodox band-aid. However, looking at the existing component logic, I guess we would already have this problem with `initializeListType`, so it's probably wrong for me to bring that up in this PR.

It also seems odd that we're doing some list-type-detection logic in the component and other logic here in the foundation, which makes me wonder if we could manage to combine the two in the foundation, but that's probably enough effort to spin out into a separate task.

Reply:
It is required for foundation to know when the list count or list type is changed. I renamed `init` to `layout` to avoid confusion and this pattern is also followed by other components when layout is changed.

Earlier foundation use to detect the list type on each event handle callback on target list item, after this change foundation does this once during initialization. It is mandatory to manually call `layout` method when list variant is changed (i.e., singlseSelection => checkbox variant - which is I think unusual case).

I had the same idea of moving component logic - `initializeListType` to foundation. But, this section of logic is replaced by props / states in other frameworks like React.

Example:

```<List selectedIndex="8" singleSelection=true />```